### PR TITLE
Refactor SDK span creation and implementation

### PR DESCRIPTION
**Pull request by MrAlias**
The span creation and configuration process is split across the tracer Start method and the startSpanInternal function, each living in different files. This adds confusion when developing. It requires the developer to remember certain parts of the configuration happen in one place or the other. Additionally, when it is determined that a span is not recorded during its creation it still is implemented by a full functioning `Span` implementation. This adds unneeded memory overhead. These changes unify the creation and configuration of a new span and add a new `nonRecordingSpan` type.

Unification of the span creation in a newSpan method on the tracer inherently associates the tracer with the span creation and centralizes all the configuration needed for a new span. This creation path is split in the newSpan method. One for a non-recording and another for a recording span.

# Benchmarks

Benchmarks performed in the `otel/sdk/trace` package.

## Before

```sh
$ go test -bench=.
go.opentelemetry.io/otel/sdk/trace.recordStackTrace(...)
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/otel/sdk/trace
cpu: Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
BenchmarkAttributesMapToKeyValue-8     	  368532	      3135 ns/op
BenchmarkStartEndSpan/AlwaysSample-8   	 1372064	       872.5 ns/op	     848 B/op	       9 allocs/op
BenchmarkStartEndSpan/NeverSample-8    	 1588966	       741.9 ns/op	     752 B/op	       8 allocs/op
BenchmarkSpanWithAttributes_4/AlwaysSample-8         	  714387	      1647 ns/op	    1584 B/op	      17 allocs/op
BenchmarkSpanWithAttributes_4/NeverSample-8          	 1367698	       877.7 ns/op	     944 B/op	       9 allocs/op
BenchmarkSpanWithAttributes_8/AlwaysSample-8         	  551848	      2054 ns/op	    2112 B/op	      23 allocs/op
BenchmarkSpanWithAttributes_8/NeverSample-8          	 1257816	       953.5 ns/op	    1136 B/op	       9 allocs/op
BenchmarkSpanWithAttributes_all/AlwaysSample-8       	  557142	      1965 ns/op	    1936 B/op	      21 allocs/op
BenchmarkSpanWithAttributes_all/NeverSample-8        	 1216846	       931.6 ns/op	    1072 B/op	       9 allocs/op
BenchmarkSpanWithAttributes_all_2x/AlwaysSample-8    	  323812	      3457 ns/op	    3236 B/op	      32 allocs/op
BenchmarkSpanWithAttributes_all_2x/NeverSample-8     	 1000000	      1177 ns/op	    1392 B/op	       9 allocs/op
BenchmarkTraceID_DotString-8                         	14290009	        80.33 ns/op
BenchmarkSpanID_DotString-8                          	19520773	        59.88 ns/op
PASS
```

## After

```sh
$ go test -bench=.
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/otel/sdk/trace
cpu: Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
BenchmarkAttributesMapToKeyValue-8     	  382208	      2932 ns/op
BenchmarkStartEndSpan/AlwaysSample-8   	 1284544	       911.1 ns/op	     848 B/op	       9 allocs/op
BenchmarkStartEndSpan/NeverSample-8    	 3394278	       341.1 ns/op	     224 B/op	       3 allocs/op
BenchmarkSpanWithAttributes_4/AlwaysSample-8         	  690414	      1576 ns/op	    1584 B/op	      17 allocs/op
BenchmarkSpanWithAttributes_4/NeverSample-8          	 2494444	       474.5 ns/op	     416 B/op	       4 allocs/op
BenchmarkSpanWithAttributes_8/AlwaysSample-8         	  495074	      2243 ns/op	    2112 B/op	      23 allocs/op
BenchmarkSpanWithAttributes_8/NeverSample-8          	 2070771	       569.0 ns/op	     608 B/op	       4 allocs/op
BenchmarkSpanWithAttributes_all/AlwaysSample-8       	  550532	      2135 ns/op	    1936 B/op	      21 allocs/op
BenchmarkSpanWithAttributes_all/NeverSample-8        	 2279611	       539.2 ns/op	     544 B/op	       4 allocs/op
BenchmarkSpanWithAttributes_all_2x/AlwaysSample-8    	  334038	      3591 ns/op	    3236 B/op	      32 allocs/op
BenchmarkSpanWithAttributes_all_2x/NeverSample-8     	 1617702	       742.9 ns/op	     864 B/op	       4 allocs/op
BenchmarkTraceID_DotString-8                         	13296470	        80.64 ns/op
BenchmarkSpanID_DotString-8                          	19072332	        62.09 ns/op
PASS
```

### CODE REVIEW

File: sdk/trace/tracer.go [hunk 1]
```diff
@@ -34,42 +34,120 @@ var _ trace.Tracer = &tracer{}
...
+	s := tr.newSpan(ctx, name, config)
+	if rw, ok := s.(ReadWriteSpan); ok && s.IsRecording() {
 		sps, _ := tr.provider.spanProcessors.Load().(spanProcessorStates)
 		for _, sp := range sps {
-			sp.sp.OnStart(ctx, span)
+			sp.sp.OnStart(ctx, rw)
 		}
 	}
+	if rtt, ok := s.(runtimeTracer); ok {
+		rtt.runtimeTrace(ctx)
```
Comment (Aneurysm9) [inline sdk/trace/tracer.go:56]: Should this be using a context containing the new span? Or at least capture the context returned here and use it when injecting the new span into a context?

File: sdk/trace/span.go [hunk 2]
```diff
@@ -204,7 +206,7 @@ func (s *span) SetStatus(code codes.Code, description string) {
 // will be overwritten with the value contained in attributes.
 //
 // If this span is not being recorded than this method does nothing.
-func (s *span) SetAttributes(attributes ...attribute.KeyValue) {
+func (s *recordingSpan) SetAttributes(attributes ...attribute.KeyValue) {
 	if !s.IsRecording() {
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply sdk/trace/span.go:210]: If the  span ends it switches to not recording, as per the specification. I left these in here for that reason.

File: sdk/trace/tracer.go (hunk 1 above)
Comment (MrAlias) [reply sdk/trace/tracer.go:56]: Yes, the returned context should be assigned. That is a bug.

**Review by tobert — COMMENTED** [review]
I was looking something up for otel-cli and happened to see this PR and read it over. I didn't see any mistakes, and having read this code on and off for the last few months, I really like how much clearer the code path is this way and this feels like a good change.

File: sdk/trace/span.go [hunk 3]
```diff
@@ -575,81 +577,65 @@ func (s *span) addChild() {
...
-	} else {
-		tid = psc.TraceID()
-		sid = provider.idGenerator.NewSpanID(ctx, tid)
-	}
+// nonRecordingSpan is a minimal implementation of the OpenTelemetry Span API
+// that wraps a SpanContext. It performs no operations other than to return
+// the wrapped SpanContext or TracerProvider that created it.
+type nonRecordingSpan struct {
+	// tracer is the SDK tracer that created this span.
+	tracer *tracer
```
Comment (jmacd) [inline sdk/trace/span.go:603]: Could you embed a `noopSpan`, then only implement `SpanContext()`?

**Comment by jmacd** [comment]
(This is a nice optimization!)

**Review by XSAM — APPROVED** [review]
No further question. Looks good. 💯

**Review by Aneurysm9 — APPROVED** [review]

File: sdk/trace/span.go (hunk 3 above)
Comment (Aneurysm9) [reply sdk/trace/span.go:603]: `noopSpan` is in the API and unexported, so it would be unavailable to use here.

Comment (jmacd) [reply sdk/trace/span.go:603]: oh, right! this is still possible, though maybe not worth it, using `SpanFromContext(context.Background())` to get one of the `noopSpan` objects as a `trace.Span`. looks like this:

https://play.golang.org/p/lBS2v7L1TWg

**Review by jmacd — APPROVED** [review]

File: sdk/trace/span.go (hunk 3 above)
Comment (MrAlias) [reply sdk/trace/span.go:603]: I think ["a little copying is better than a little dependency"](https://go-proverbs.github.io/) applies here. It is probably a good exercise to take a fresh look at the span types and see if we could restructure them to be universal to the project, but it seems that would be a good independent change for its own PR.

File: sdk/trace/tracer.go [hunk 4]
```diff
@@ -34,42 +34,120 @@ var _ trace.Tracer = &tracer{}
...
+		TraceID:    tid,
+		SpanID:     sid,
+		TraceState: samplingResult.Tracestate,
+	}
+	if isSampled(samplingResult) {
+		scc.TraceFlags = psc.TraceFlags() | trace.FlagsSampled
+	} else {
+		scc.TraceFlags = psc.TraceFlags() &^ trace.FlagsSampled
+	}
+	sc := trace.NewSpanContext(scc)
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply sdk/trace/tracer.go:110]: I'm not following. The quoted code section is creating a `SpanContext`, but it is also generating appropriate span and trace IDs and determining the sampling result. The `SpanContext` that is created is indeed stored in the produced span, but I'm not seeing where in the `Start` method it is being extracted from the span.

I'm guessing I'm missing something here. I'd like to better understand what you are suggesting. Can you provide an example?

[the remark this answers is not in the document]
Comment (MadVikingGod) [reply sdk/trace/tracer.go:110]: So I must have been chasing ghost when reviewing this.  I still think extracting this spancontext creation would be a good idea for clarity.  But, I didn't intend to hold up the merge of this.

**Review by MadVikingGod — APPROVED** [review]

**Pull request merged**